velox/expression/Expr.cpp

```diff
@@ -259,6 +289,110 @@ bool isFlat(const BaseVector& vector) {
 
 } // namespace
 
+bool MutableRemainingRows::deselectNonErrorNulls(
+    const VectorPtr& result,
+    EvalCtx::ErrorVectorPtr& errors) {
+  const uint64_t* flatNulls = nullptr;
+  if (result->mayHaveNulls()) {
+    decoded_.get()->decode(*result, rows());
+    flatNulls = decoded_.get()->nulls();
+  }
+  // A null with no error deselects the row and clears a possible error. An
+  // error adds itself to argument errors.
+  if (context_.errors()) {
+    // There are new errors.
+    auto& remainingRows = mutableRows();
+    context_.ensureErrorsVectorSize(*context_.errorsPtr(), rows().end());
+    auto newErrors = context_.errors();
+    if (flatNulls) {
+      // There are both nulls and errors. Only a null with no error removes a
+      // row.
+      auto errorNulls = newErrors->rawNulls();
+      auto rowBits = mutableRows().asMutableRange().bits();
+      auto nwords = bits::nwords(rows().end());
+      for (auto i = 0; i < nwords; ++i) {
+        auto nullNoError = errorNulls ? flatNulls[i] | errorNulls[i] : 0;
+        rowBits[i] &= nullNoError;
+        if (errors) {
+          errors->mutableRawNulls()[i] &= nullNoError;
+        }
+      }
+      mutableRows().updateBounds();
+    }
+    context_.moveAppendErrors(errors);
+  } else {
+    // There are no new errors, a null deselects a row and clears an error.
+    if (flatNulls) {
+      if (errors) {
+        errors->addNulls(flatNulls, rows());
+      }
+      deselectNulls(flatNulls);
+    }
+  }
+  return rows().hasSelections();
+}
+
+template <typename EvalArg>
+bool Expr::evalArgsDefaultNulls(
+    MutableRemainingRows& rows,
+    EvalArg evalArg,
+    EvalCtx& context,
+    VectorPtr& result) {
+  EvalCtx::ErrorVectorPtr argumentErrors;
+  EvalCtx::ErrorVectorPtr originalErrors;
+  // Store pre-existing errors locally and clear them from
+  // 'context'. We distinguish between argument errors and
+  // pre-existing ones.
+  if (context.errors()) {
+    context.swapErrors(originalErrors);
+  }
+
+  inputValues_.resize(inputs_.size());
+  {
+    ScopedVarSetter throwErrors(
+        context.mutableThrowOnError(), throwArgumentErrors(context));
+
+    for (int32_t i = 0; i < inputs_.size(); ++i) {
+      evalArg(i);
+      if (!rows.deselectNonErrorNulls(inputValues_[i], argumentErrors)) {
+        break;
+      }
+    }
+  }
+
+  mergeOrThrowArgumentErrors(
+      rows.originalRows(), originalErrors, argumentErrors, context);
+
+  if (!rows.deselectErrors()) {
```

Comment:
nit: feel free to ignore this comment:
this is common to both evalArgsDefaultNulls and evalArgsWithNulls, maybe we can do this outside these eval functions? This way we can get rid of storing originalRows() in MutableRemainingRows as well since they will be no longer needed.